### ISSUE TITLE
Undefined CUDA symbols

### ISSUE THREAD

**Issue by spakin** [diku-dk/futhark#1801]
Futhark had been working fine, but I recently upgraded CUDA, and now I'm getting missing-symbol errors when I compile a program with `futhark cuda`:
```console
$ cat trivial.fut 
def main : i32 = 2 + 3
$ env CFLAGS='-O -std=c99 -I/usr/local/cuda/targets/x86_64-linux/include -L/usr/local/cuda/targets/x86_64-linux/lib' futhark cuda trivial.fut 
$ ./trivial 
./trivial: NVRTC compilation failed.

futhark-cuda(1668): error: identifier "mulhi" is undefined

futhark-cuda(1672): error: identifier "mul64hi" is undefined

2 errors detected in the compilation of "futhark-cuda".
```
Here's what I currently have installed (Ubuntu 22.04.1):
```console
$ dpkg -l cuda
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
ii  cuda           12.0.0-1     amd64        CUDA meta-package
$ futhark --version
Futhark 0.23.0 (prerelease - include info below when reporting bugs)
git: c9b8dcbd06f437ed0320f7bdf518b59d8055983b
Copyright (C) DIKU, University of Copenhagen, released under the ISC license.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

I tried replacing the `cuda` package with `nvidia-cuda-toolkit`, but that just made things worse:
```console
$ ./trivial 
./trivial: trivial.c:6626: CUDA call
  cuInit(0)
failed with error code 803 (system has unsupported display driver / cuda driver combination)
```

Any suggestions would be appreciated.

**Comment by athas** [comment]
Interesting.  I wonder if CUDA 12 changed the name of the `mulhi` function (it seems to be called `__mulhi` in the documentation).  I don't have a CUDA 12 system at hand, but I will investigate.

If you are able to compile the Futhark compiler, try changing these two lines by prefixing two underscores to the function name, and see if that makes a difference:

https://github.com/diku-dk/futhark/blob/c9b8dcbd06f437ed0320f7bdf518b59d8055983b/rts/c/scalar.h#L1325

https://github.com/diku-dk/futhark/blob/c9b8dcbd06f437ed0320f7bdf518b59d8055983b/rts/c/scalar.h#L1329

**Comment by spakin** [comment]
Thanks for the prompt response.  I managed to compile Futhark and make the changes you recommended.  I now receive a different error message at run time:

```console
$ ./trivial 
./trivial: NVRTC compilation failed.

futhark-cuda(1668): error: identifier "__return" is undefined

1 error detected in the compilation of "futhark-cuda".
```

**Comment by athas** [comment]
Did you add the underscores to the `return` keyword?  You should add them to `mulhi`, so the line reads:

```
 return __mulhi(a, b); 
```

**Comment by spakin** [comment]
Argh!  That's indeed what I did on line 1325, although I got it correct on line 1329.  (My brain knew where the underscores were supposed to go, but my fingers weren't paying attention. :frowning: )

`futhark cuda` now works with CUDA 12—thanks!

**Comment by athas** [comment]
Alright, I'll test whether this also works on older CUDA versions and modify the compiler itself if so.

**Comment by athas** [comment]
Unfortunately, while this compiles on older CUDA versions, it produces incorrect results (the words are entirely filled with `1` bits).  I wonder what is going on here.

**Comment by spakin** [comment]
I confirm that Futhark c146344fcf5d9f5009a017ff243b0a5bf4fed80f works fine with CUDA 12.  (I actually tested slightly later version ae3df379c0081fa40a8c2a02e8b3a56dca17a15a).  Thanks for fixing the code!

**Comment by athas** [comment]
Great!  The fix is also part of the latest release (0.22.6).
